onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -0,0 +1,481 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+#include "ml_common.h"
+#include <math.h>
+
+namespace onnxruntime {
+namespace ml {
+
+struct TreeNodeElementId {
+  int tree_id;
+  int node_id;
+  inline bool operator==(const TreeNodeElementId& xyz) const {
+    return (tree_id == xyz.tree_id) && (node_id == xyz.node_id);
+  }
+  inline bool operator<(const TreeNodeElementId& xyz) const {
+    return ((tree_id < xyz.tree_id) || (tree_id == xyz.tree_id && node_id < xyz.node_id));
+  }
+};
+
+template <typename T>
+struct SparseValue {
+  int64_t i;
+  T value;
+};
+
+enum MissingTrack {
+  NONE,
+  TRUE,
+  FALSE
+};
+
+template <typename T>
+struct TreeNodeElement {
+  TreeNodeElementId id;
+  int feature_id;
+  T value;
+  T hitrates;
+  NODE_MODE mode;
+  TreeNodeElement<T>* truenode;
+  TreeNodeElement<T>* falsenode;
+  MissingTrack missing_tracks;
+  std::vector<SparseValue<T>> weights;
+
+  bool is_not_leave;
+  bool is_missing_track_true;
+};
+
+template <typename ITYPE, typename OTYPE>
+class _Aggregator {
+ protected:
+  size_t n_trees_;
+  int64_t n_targets_or_classes_;
+  POST_EVAL_TRANSFORM post_transform_;
+  const std::vector<OTYPE>* base_values_;
+  OTYPE origin_;
+  bool use_base_values_;
+
+ public:
+  inline _Aggregator(size_t n_trees,
+                     const int64_t& n_targets_or_classes,
+                     POST_EVAL_TRANSFORM post_transform,
+                     const std::vector<OTYPE>* base_values) : n_trees_(n_trees), n_targets_or_classes_(n_targets_or_classes), post_transform_(post_transform), base_values_(base_values) {
+    origin_ = base_values_->size() == 1 ? (*base_values_)[0] : 0.f;
+    use_base_values_ = base_values_->size() == static_cast<size_t>(n_targets_or_classes_);
+  }
+
+  // 1 output
+
+  inline void ProcessTreeNodePrediction1(OTYPE* /*predictions*/, TreeNodeElement<OTYPE>* /*rooOTYPE*/,
+                                         unsigned char* /*has_predictions*/) const {}
+
+  inline void MergePrediction1(OTYPE* /*predictions*/, unsigned char* /*has_predictions*/,
+                               OTYPE* /*predictions2*/, unsigned char* /*has_predictions2*/) const {}
+
+  inline void FinalizeScores1(OTYPE* Z, OTYPE& val,
+                              unsigned char& has_scores,
+                              int64_t*) const {
```

Comment:
Does this need to be a non-const reference or could it just be passed by value?

Reply:
All parameters can be modified by the fonction.